packages/iota/src/validation/features/features.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+import { HexHelper } from "@iota/util.js";
+import { MAX_TAG_LENGTH } from "../../binary/payloads/taggedDataPayload";
+import type { FeatureTypes } from "../../models/features/featureTypes";
+import { ISSUER_FEATURE_TYPE } from "../../models/features/IIssuerFeature";
+import { IMetadataFeature, METADATA_FEATURE_TYPE } from "../../models/features/IMetadataFeature";
+import { SENDER_FEATURE_TYPE } from "../../models/features/ISenderFeature";
+import { ITagFeature, TAG_FEATURE_TYPE } from "../../models/features/ITagFeature";
+import type { ITypeBase } from "../../models/ITypeBase";
+import { ValidationHelper } from "../../utils/validationHelper";
+import { validateAddress } from "../addresses/addresses";
+import { IValidationResult, mergeValidationResults, failValidation } from "../result";
+
+/**
+ * The maximum length of a metadata binary representation.
+ */
+export const MAX_METADATA_LENGTH: number = 8192;
+
+/**
+ * Validate output features.
+ * @param object The object to validate.
+ * @returns The validation result.
+ */
+// Each output must not contain more than one feature of
+// each type and not all feature types are supported for each output type.
+export function validateFeatures(object: FeatureTypes[]): IValidationResult {
+    const results: IValidationResult[] = [];
+
+    if (object.length > 4) {
+        results.push({
+            isValid: false,
+            errors: ["Max number of features exceeded."]
+        });
+    }
+
+    const distinctFeatureTypes = new Set(object.map(feature => feature.type));
```

Comment:
Push the code that constructs the set into `validateDistinct`. Using a set is an implementation detail of that function, so if should be inside.
Then we will have `validateDistinct(object.map(f => f.type), object.length, ...)`